### DISCUSSION TITLE
How many hosts for proper testing infrastructure?

### DISCUSSION THREAD

**Discussion by az-2022** [TPC-Council/HammerDB#203] [Q&A] [answered]
Hi.  How many hosts and roles are required for enterprise level testing infrastructure?   The "Test Network Configuration" section in the documentation discusses three components or roles:

SUT database Server
Load Generation Server
Administrator PC

Is it advised to split the HammerDB interface from the load generation (virtual users) onto multiple hosts?  If so, how to accomplish that?  

I only know that I was able to install HammerDB 4 on my corporate Windows PC and successfully run the "Quick Start" scenario against a SQL 2017 SUT.

Thanks.

**Comment by sm-shaw** [comment, marked as answer]
Hi. Great question again. This very much depends on the database being tested. For the commercial databases these can run at very high transaction rates and use all of the CPU on database server, therefore if you put HammerDB on the same server then the overall throughput is likely to be lower and of course it is not as straightforward to separate the HammerDB use of CPU from the database. For the open source databases these don't tend to be able to use all of the CPU on an up to date server and therefore you may find putting HammerDB on the database server gives higher performance.  They key thing is to be consistent and if you are making comparisons to test in the same way. 
Fortunately HammerDB is designed from the ground up to be very, very efficient and lightweight in its footprint as a load generator so you may not need multiple load generators.  A typical 2 socket server or PC an drive high levels of load on a database. (For SQL Server if testing over the network keep and eye on the network bandwidth as high throughput can exceed for example a gigabit network card on a PC.) If you do wish to use multiple load generators at the same time there is a feature called Primary/Replica for this https://www.hammerdb.com/docs/ch08.html so you can run multiple load generators simultaneously from systems over the network. I would suggest initially running some timed tests and looking at the statistics on your load generator and database server. The most common error is creating too many virtual users to start with, initially start with 1 then 2, 4 etc and build up. Once you have a stable environment use Autopilot to automate this workload and plot your performance https://www.hammerdb.com/docs/ch05.html this will help you find a peak in throughput where you can investigate any limitations on performance. also consider the CPU metrics https://www.hammerdb.com/docs/ch07s03.html - on Windows pay attention to the distribution of the workload across multiple processor groups.
